plugins/externalwrench/src/ExternalWrench.cc

```diff
@@ -115,28 +118,133 @@ bool ExternalWrench::setWrench(physics::ModelPtr& _model,yarp::os::Bottle& cmd)
 
     if(getLink())
     {
-        wrench.force[0]  =  cmd.get(1).asDouble();
-        wrench.force[1]  =  cmd.get(2).asDouble();
-        wrench.force[2]  =  cmd.get(3).asDouble();
+        // Get wrench duration
+        wrench.duration  = cmd.get(7).asDouble();
 
-        wrench.torque[0] = cmd.get(4).asDouble();
-        wrench.torque[1] = cmd.get(5).asDouble();
-        wrench.torque[2] = cmd.get(6).asDouble();
+        if (wrenchSmoothing) {
+            wrenchSmoothingFlag = true;
+            bool wrenchSmoothingStatus = smoothWrench(cmd, simulationUpdatePeriod);
 
-        wrench.duration  = cmd.get(7).asDouble();
+            if (wrenchSmoothingStatus) {
+
+                // Get the first elements of the smoothed wrench
+                yarp::sig::Vector initSmoothedWrench = wrench.smoothedWrenchVec.at(timeStepIndex);
+
+                wrench.force[0] = initSmoothedWrench[0];
+                wrench.force[1] = initSmoothedWrench[1];
+                wrench.force[2] = initSmoothedWrench[2];
+
+                wrench.torque[0] = initSmoothedWrench[3];
+                wrench.torque[1] = initSmoothedWrench[4];
+                wrench.torque[2] = initSmoothedWrench[5];
+
+                return true;
+            }
+            else return  false;
+        }
+        else {
+
+            wrench.force[0]  =  cmd.get(1).asDouble();
+            wrench.force[1]  =  cmd.get(2).asDouble();
+            wrench.force[2]  =  cmd.get(3).asDouble();
+
+            wrench.torque[0] = cmd.get(4).asDouble();
+            wrench.torque[1] = cmd.get(5).asDouble();
+            wrench.torque[2] = cmd.get(6).asDouble();
+
+            return true;
+        }
 
-        return true;
     }
     else return false;
 }
 
+bool ExternalWrench::smoothWrench(const yarp::os::Bottle& cmd, const double& simulationUpdatePeriod)
+{
+    // Clear smoothed wrenches vector
+    wrench.smoothedWrenchVec.clear();
+
+    double duration = cmd.get(7).asDouble();
+
+    // Compute time steps
+    steps = duration/simulationUpdatePeriod;
+
+    // Get original wrench
+    yarp::sig::Vector originalWrench;
+    originalWrench.resize(6,0);
+
+    originalWrench[0]  =  cmd.get(1).asDouble();
+    originalWrench[1]  =  cmd.get(2).asDouble();
+    originalWrench[2]  =  cmd.get(3).asDouble();
+    originalWrench[3]  =  cmd.get(4).asDouble();
+    originalWrench[4]  =  cmd.get(5).asDouble();
+    originalWrench[5]  =  cmd.get(6).asDouble();
+
+    double time = 0;
+    std::vector<double> smoothingCoefficients;
+
+    for (int timeStep = 0; timeStep <= steps; timeStep++) {
+        // Second derivative of minimum jerk trajectory
```

Comment:
Could you please document this in detail?

Reply:
Done https://github.com/Yeshasvitvs/gazebo-yarp-plugins/commit/b33cfcfca4a80c85ea300ba6cc3ec19a31ba8f8e